scipy/optimize/_optimize.py

```diff
@@ -3477,12 +3489,23 @@ def brute(func, ranges, args=(), Ns=20, full_output=0, finish=fmin,
     for k in range(N):
         if type(lrange[k]) is not type(slice(None)):
             if len(lrange[k]) < 3:
-                lrange[k] = tuple(lrange[k]) + (complex(Ns),)
+                if integrality[k]:
+                    ub = np.floor(lrange[k][1])
+                    lb = min(np.ceil(lrange[k][0]), ub)
+                    lrange[k] = lb, ub
+                    n_ints_between_bounds = lrange[k][1] - lrange[k][0] + 1
+                    n_points = min(Ns, n_ints_between_bounds)
+                else:
+                    n_points = Ns
+                lrange[k] = tuple(lrange[k]) + (complex(n_points),)
             lrange[k] = slice(*lrange[k])
     if (N == 1):
         lrange = lrange[0]
 
     grid = np.mgrid[lrange]
+    for k in range(N):
+        if integrality[k]:
+            grid[k, ...] = np.round(grid[k, ...])
```

Comment:
I think this is going to result in repeat values being present, e.g.:
```
np.round(np.mgrid[slice(0, 10, 21j)])
```
gives
```
array([ 0.,  0.,  1.,  2.,  2.,  2.,  3.,  4.,  4.,  4.,  5.,  6.,  6.,
        6.,  7.,  8.,  8.,  8.,  9., 10., 10.])
```
i.e. there are 3 fours, and that's going to result in a large amount of un-necessary computation (21 computations instead of 11). In multidimensional work that'll be a big performance barrier.

Reply:
Is it possible to do: `grid = np.unique(grid, axis=k+1)` here?

Reply:
If the user provides lower, upper bounds, this is prevented by converting to a slice with points at least 1 apart.
```
                    n_ints_between_bounds = lrange[k][1] - lrange[k][0] + 1
                    n_points = min(Ns, n_ints_between_bounds)
```
If the user provides a slice, yes, they can force `brute` to perform a lot of redundant calculation.

I did think about using `np.unique` while writing this, but tried to avoid it with the logic for lower, upper bounds, and I chose not to get in the way of slices.

Reply:
If `integrality` is ignored for slices then the uniqueness issue goes away, as does evaluating outside the bounds.

If we keep `integrality` for slices then I think we should be checking for bounds, enforcing uniqueness, presence of at least one integer.

Reply:
I think this is addressed by [this](https://github.com/scipy/scipy/pull/15389#discussion_r800086755). Now, arrays generated by slices take on only integer values, too.